fedimint-client/src/module/mod.rs

```diff
@@ -725,38 +645,69 @@ pub trait IClientModule: Debug {
 
     async fn backup(&self, module_instance_id: ModuleInstanceId)
         -> anyhow::Result<DynModuleBackup>;
+}
 
-    fn supports_being_primary(&self) -> bool;
-
+#[apply(async_trait_maybe_send!)]
+pub trait BalanceProvider: 'static + MaybeSend + MaybeSync {
+    /// Creates an input of **at least** a given `min_amount` from the holdings
+    /// managed by the module.
+    ///
+    /// If successful it returns:
+    /// * A set of private keys belonging to the input for signing the
+    ///   transaction
+    /// * The input of **at least** `min_amount`, the actual amount might be
+    ///   larger, the caller has to handle this case and possibly generate
+    ///   change using `create_change_output`.
+    /// * A closure that generates states belonging to the input. This closure
+    ///   takes the transaction id of the transaction in which the input was
+    ///   used and the input index as input since these cannot be known at time
+    ///   of calling `create_funding_input` and have to be injected later.
+    ///
+    /// The function returns an error if the client's funds are not sufficient
+    /// to create the requested input.
     async fn create_sufficient_input(
         &self,
-        module_instance: ModuleInstanceId,
+        client: &Client,
```

Comment:
Modules already have access to `Client` via `ClientCtx` passed in `args` in `init`.

Reply:
this is not meant to be implemented for Module type
https://github.com/fedimint/fedimint/pull/5079#discussion_r1578481967